redash/query_runner/json_ds.py

```diff
@@ -0,0 +1,222 @@
+    while len(path_parts) > 0:
+        current_path = path_parts.pop()
+        if current_path in response:
+            response = response[current_path]
+        else:
+            raise Exception(
+                "Couldn't find path {} in response.".format(path))
+
+    return response
+
+
+def _normalize_json(data, path):
+    data = _apply_path_search(data, path)
+
+    if isinstance(data, dict):
+        data = [data, ]
+
+    return data
+
+
+def _sort_columns_with_fields(columns, fields):
+    if fields:
+        columns = compact(
+            [_get_column_by_name(columns, field) for field in fields])
+
+    return columns
+
+
+# TODO: merge the logic here with the one in MongoDB's queyr runner
+def parse_json(data, path, fields):
+    data = _normalize_json(data, path)
+
+    rows = []
+    columns = []
+
+    for row in data:
+        parsed_row = {}
+
+        for key in row:
+            if isinstance(row[key], dict):
+                for inner_key in row[key]:
+                    column_name = u'{}.{}'.format(key, inner_key)
+                    if fields and key not in fields and column_name not in fields:
+                        continue
+
+                    value = row[key][inner_key]
+                    add_column(columns, column_name, _get_type(value))
+                    parsed_row[column_name] = value
+            else:
+                if fields and key not in fields:
+                    continue
+
+                value = row[key]
+                add_column(columns, key, _get_type(value))
+                parsed_row[key] = row[key]
+
+        rows.append(parsed_row)
+
+    columns = _sort_columns_with_fields(columns, fields)
+
+    return {'rows': rows, 'columns': columns}
+
+
+class JSON(BaseHTTPQueryRunner):
+    requires_url = False
+
+    @classmethod
+    def configuration_schema(cls):
+        return {
+            'type': 'object',
+            'properties': {
+                'username': {
+                    'type': 'string',
+                    'title': cls.username_title,
+                },
+                'password': {
+                    'type': 'string',
+                    'title': cls.password_title,
+                },
+            },
+            'secret': ['password'],
+            'order': ['username', 'password']
+        }
+
+    @classmethod
+    def annotate_query(cls):
+        return False
+
+    def __init__(self, configuration):
+        super(JSON, self).__init__(configuration)
+        self.syntax = 'yaml'
+
+    def test_connection(self):
+        pass
+
+    def run_query(self, query, user):
+        try:
+            query = parse_query(query)
+
+            if not isinstance(query, dict):
+                raise QueryParseError(
+                    "Query should be a YAML object describing the URL to query.")
+
+            if 'url' not in query:
+                raise QueryParseError("Query must include 'url' option.")
+
+            if is_private_address(query['url']):
```

Comment:
Why should we avoid querying private url/ip ?

Reply:
Because in some environments being able to query private addresses might allow you to poke around things you shouldn't (like AWS metadata of the instance hosting Redash). We can make this optional if needed.

Reply:
I think it's a good idea to have an option to disable it